File: python/1792.maximum-average-pass-ratio/solution.py
# Created by none at 2024/12/15 14:22
# leetgo: 1.4.9
# https://leetcode.com/problems/maximum-average-pass-ratio/

from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, cmp_to_key, lru_cache, reduce
from heapq import heapify, heappop, heappush, heappushpop, heapreplace
from itertools import accumulate, chain, count, pairwise, zip_longest
from math import ceil, comb, floor, gcd, inf, isqrt, lcm, log2, perm, sqrt
from operator import xor
from pprint import pprint
from string import ascii_lowercase
from typing import *
from typing import List, Optional

from leetgo_py import *

# @lc code=begin


class PassRatio:
    def __init__(self, p: int, total: int):
        self.p = p
        self.total = total

    def __str__(self):
        return f"{self.p} {self.total}"

    def __lt__(self, other: "PassRatio"):
        # python heap is min heap
        # here we want (a+1)/(b+1) - (a/b) maxium
        return (self.total - self.p) * (other.total * (other.total + 1)) >= (
            other.total - other.p
        ) * (self.total * (self.total + 1))


class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        q = []
        for p, t in classes:
            heappush(q, PassRatio(p, t))
        for _ in range(extraStudents):
            pr = heappop(q)
            heappush(q, PassRatio(pr.p + 1, pr.total + 1))
        return sum(pr.p / pr.total for pr in q) / len(q)


# @lc code=end

if __name__ == "__main__":
    classes: List[List[int]] = deserialize("List[List[int]]", read_line())
    extraStudents: int = deserialize("int", read_line())
    ans = Solution().maxAverageRatio(classes, extraStudents)
    print("\noutput:", serialize(ans, "double"))

File: go/1792.maximum-average-pass-ratio/solution.go
// Created by none at 2025/09/01 12:18
// leetgo: dev
// https://leetcode.com/problems/maximum-average-pass-ratio/

package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin
type Item struct {
	delta float64
	x, y  int
}

type Heap []Item

func (h Heap) Len() int {
	return len(h)
}

func (h Heap) Less(i, j int) bool {
	return h[i].delta > h[j].delta
}

func (h *Heap) Push(item any) {
	*h = append((*h), item.(Item))
}

func (h *Heap) Pop() any {
	item := (*h)[h.Len()-1]
	*h = (*h)[:h.Len()-1]
	return item
}

func (h *Heap) Swap(i int, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

func maxAverageRatio(classes [][]int, extraStudents int) float64 {
	n := len(classes)

	h := make(Heap, n)
	for i, c := range classes {
		x, y := c[0], c[1]
		h[i] = Item{float64(y-x) / float64(y+1) / float64(y), x, y}
	}
	heap.Init(&h)

	for range extraStudents {
		h[0].x++
		h[0].y++
		x, y := h[0].x, h[0].y
		h[0].delta = float64(y-x) / float64(y+1) / float64(y)
		heap.Fix(&h, 0)
	}

	res := 0.0
	for _, item := range h {
		res += float64(item.x) / float64(item.y)
	}
	return res / float64(n)
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	classes := Deserialize[[][]int](ReadLine(stdin))
	extraStudents := Deserialize[int](ReadLine(stdin))
	ans := maxAverageRatio(classes, extraStudents)

	fmt.Println("\noutput:", Serialize(ans))
}
